Introduction of RxData to handle receive data

Toreil opened the pull request:
The current implementation of the way that the data is received has certain limitations and feels slow since the data is only processed at the end. 

In this PR we introduce the RxData object to handle storage and processing of the receive data. An RxData object is created for each ADC event defined in the sequence provider, describing all of the properties of that ADC event, such as number of (decimated) data points, dwell time, phase offset etc. The sequence provider returns a list of all RxData objects to the acquisition controller which in turn passes that information to the rx_device which then acquires the raw data and stores it in the RxData object. Data processing is then done for each RxData object separately, which means that each RxData object can be demodulated with a different frequency, decimated with a different factor etc.

To improve the performance of the data processing the RxData can be processed in a multi-threaded or a multi-processing approach. The multi-processing approach is specifically designed to process the data while it is coming in, meaning that the processed data is available essentially immediately after the acquisition has been finished.

- [x] Test basic implementation, without real time data processing
~~Test real time data processing~~
~~Test memory enhancement through discarding of unprocessed data~~
- [x] Add unit tests for the new `RxData`
- Tests are included in the existing tests

Toreil commented:
Currently the code for acquiring and processing the data is running but it has not been tested on an image acquisition yet.

Toreil commented:
Real time data processing will be included in a later update

schote reviewed (commented):
Thank you @Toreil, I think this is a very important update

Toreil commented on src/console/pulseq_interpreter/sequence_provider.py:
```diff
@@ -569,27 +564,47 @@ def unroll_sequence(self, parameter: AcquisitionParameter) -> UnrolledSequence:
                 # Add ADC gate to X gradient
                 _seq[adc_start + 1:adc_end + 1:4] = _seq[adc_start + 1:adc_end + 1:4] | adc_waveform
 
+                _rx_data.append(RxData(index=adc_count,
```
I've done this now but I dont know if there is a nice way to ensure that a label belongs to an ADC event, at the moment it's assumed that the each adc event has the same number of labels but this may not be true.

schote replied:
On page 17 of this pulseq [document](https://pulseq.github.io/specification.pdf) it is written, that all counters and flags are automatically initialized to zero. So we should always get all labels, even though some of them are not used, no?

Toreil replied:
Reading labels directly from the block rather than using evaluate_labels works better and ensures that the correct label is linked to the correct rx_data event. This has now been implemented.

Toreil commented on src/console/spcm_control/rx_device.py:
```diff
@@ -83,8 +83,16 @@ def __init__(
-        self.rx_scaling = [amp / (2**16) for amp in self.max_amplitude]
+        self.rx_scaling = [amp / (2**15) for amp in self.max_amplitude]
+
+    def dict(self) -> dict:
```
This was unintentionally put back in. I've been debating whether or not the scaling factor should be stored in RxData, on one hand it would mean that all the info to go from raw data to processed data is in the data object, on the other hand it has to be the same for each RxData object in a scan session so it's redundant information.

Toreil commented:
I just tested it and it seems to duplicate the labels of the previous event which is a problem. Maybe it’s possible to add some extra checks, I’ll try that asap.On 18 Jun 2025, at 22:40, David Schote ***@***.***> wrote:﻿
@schote commented on this pull request.



In src/console/pulseq_interpreter/sequence_provider.py:
> @@ -569,27 +564,47 @@ def unroll_sequence(self, parameter: AcquisitionParameter) -> UnrolledSequence:
                 # Add ADC gate to X gradient
                 _seq[adc_start + 1:adc_end + 1:4] = _seq[adc_start + 1:adc_end + 1:4] | adc_waveform
 
+                _rx_data.append(RxData(index=adc_count,

On page 17 of this pulseq document it is written, that all counters and flags are automatically initialized to zero. So we should always get all labels, even though some of them are not used, no?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

Toreil commented on src/console/interfaces/rx_data.py:
```diff
@@ -0,0 +1,183 @@
+
+    def add_item(self, item: RxData, raw_data: np.ndarray, larmor_freq: float) -> None:
+        """Add and immediately submit item for processing"""
+        result = self.pool.apply_async(RxData.set_and_process_data, (item,raw_data, larmor_freq))
```
The shared memory/multiprocessing features will be implemented in a later update

Toreil commented on src/console/interfaces/rx_data.py:
```diff
@@ -0,0 +1,197 @@
+
+
+@dataclass
+class RxData:
```
Done

Toreil commented on src/console/spcm_control/rx_device.py:
```diff
@@ -293,7 +315,7 @@ def _gated_timestamps_stream(self):
         available_data_bytes = sp.int32(0)
         available_data_position = sp.int32(0)
-        total_gates = 0
+        self.total_gates = 0
```
It's needed to check for when all the ADC gates/events have been read by the RxCard by the acquisition controller, in the old receive approach this was done by checking the length of rx_data:
while (num_gates := len(self.rx_card.rx_data)) < self.sequence.adc_count or num_gates == 0:

But since the list of RxData events if fully populated this had to be changed to a different check which I did here. I guess it could be changed to a boolean indicating that data collection is complete but there has to be something that the acquisition_controller can check. I'm open to suggestions here.

Toreil replied:
Thanks for this code, I've implemented it in the latest commit.

Toreil commented on src/console/spcm_control/acquisition_control.py:
```diff
@@ -274,74 +290,28 @@ def run(self) -> AcquisitionData:
-            else:
-                self._raw.append(data[None, ...])
+        # Scale the data
+        for rx_data in self.receive_data[-1]:
```
I considered this too, the only thing that has made me hesitate is that it's redundant information since all RxData objects have to have the same scaling. I'll put it in for completeness but its probably better just to have it in the meta data for the sequence

Toreil commented on src/console/interfaces/rx_data.py:
```diff
@@ -0,0 +1,183 @@
+        self.process_data(larmor_freq=larmor_freq, store_unprocessed=store_unprocessed)
+
+
+class MultiprocessingProcessor:
```
Changed the threading approach to use the `ThreadPoolExecutor `and moved it to the `AcquisitionControl` class

Toreil commented on src/console/interfaces/rx_data.py:
```diff
@@ -0,0 +1,183 @@
+    def __post_init__(self) -> None:
+        """Post init method to calculate the decimation factor."""
+        self.decimation_factor = round(self.dwell_time/self.dwell_time_raw)
+
```
Decimation and demodulation are distinct processing steps that the user may want to perform separately/independently which is why I divide it up.

Toreil commented on examples/tse_3d.py:
```diff
@@ -10,39 +12,52 @@
+
+for idx_avg in range(np.size(ksp, 0)):
+    for idx_coil in range(np.size(ksp, 1)):
+        img[idx_avg, idx_coil, ...] = np.fft.fftshift(np.fft.fftn(np.fft.fftshift(ksp[idx_avg, idx_coil, ...])))
```
Agreed, I will modify this but want to do a phantom validation to ensure it's all done properly.

Toreil commented on src/console/spcm_control/rx_device.py:
```diff
@@ -413,14 +426,21 @@ def _gated_timestamps_stream(self):
+                        self.rx_data[self.total_gates].raw_data = gate_data.reshape((self.num_channels.value,
+                                                                                gate_sample),
+                                                                                order="F")
+                        self.rx_data[self.total_gates].scaling_factor = self.rx_scaling[:self.num_channels.value]
```
I think it's better to leave it untouched since we do call it raw_data, the moment you apply scaling it's no longer that. I also want to add some checks in the future for ADC overload or on the other end, not filling the ADC enough and this is more obvious to do is still in it's raw integer state.

schote replied:
Convinced

schote commented on examples/tse_3d.py:
```diff
@@ -10,39 +12,52 @@
 # Execute the sequence and sort kspace array
 acq_data: AcquisitionData = acq.run()
-ksp = tse_3d.sort_kspace(acq_data.raw, seq).squeeze()
+ksp = tse_3d.sort_kspace(acq_data.receive_data, acq.seq_provider)
```
I would not be concerned about the first point and I really like the second one.
Having the necessary labels extracted from the sequence would allow to sort the data without the sequence.
If everything is labeled properly, it should also be possible to generate the ISMRMRD file from the RxData (without the need of passing a header file).
Alternatively, it should also be possible to provide an ISMRMRD header, if the sequence is not labeled.

schote commented on examples/tse_3d.py:
```diff
@@ -10,39 +12,52 @@
+        img[idx_avg, idx_coil, ...] = np.fft.fftshift(np.fft.fftn(np.fft.fftshift(ksp[idx_avg, idx_coil, ...])))
+
+img = img.squeeze()
+ksp = ksp.squeeze()
```
I mean if you don't you squeeze here, the data should also have the dimensions `[average, coil, enc1, enc2, enc3]` shouldn't it? You could do `ksp = ksp[0, 0, ...]` and a squeeze afterwards to prevent the if-clause?

Toreil replied:
Wow, brain fart! Yes of course, not even a squeeze necessary anymore.

Toreil commented:
> Otherwise would you define the averages in the sequence?

Yeah, this would also allow you to do phase cycling in the averaging which helps eliminate RF related image artifacts but obviously increases memory consumption substantially

>  I also thought about storing the averages in the “RxData”. But that might make it more difficult to start processing at runtime.

We could append the list with another copy of RxData and add a flag for average number (for accuracy this should probably be distinct from the Pulseq flag for average), this keeps the processing of the data the same, it would just require taking some care with indexing

schote commented:
> > Otherwise would you define the averages in the sequence?
> 
> Yeah, this would also allow you to do phase cycling in the averaging which helps eliminate RF related image artifacts but obviously increases memory consumption substantially
> 
> > I also thought about storing the averages in the “RxData”. But that might make it more difficult to start processing at runtime.
> 
> We could append the list with another copy of RxData and add a flag for average number (for accuracy this should probably be distinct from the Pulseq flag for average), this keeps the processing of the data the same, it would just require taking some care with indexing

Until we overcome the constraints on the sequence duration, copying the `RxData` for a given number of repetitions is a good idea. I would use a counter for the repetions as we could employ it directly for the generation of an ISMRMRD file.

schote commented:
> > > > Otherwise would you define the averages in the sequence?
> 
> > > 
> 
> > > 
> 
> > > Yeah, this would also allow you to do phase cycling in the averaging which helps eliminate RF related image artifacts but obviously increases memory consumption substantially
> 
> > > > I also thought about storing the averages in the “RxData”. But that might make it more difficult to start processing at runtime.
> 
> > > 
> 
> > > 
> 
> > > We could append the list with another copy of RxData and add a flag for average number (for accuracy this should probably be distinct from the Pulseq flag for average), this keeps the processing of the data the same, it would just require taking some care with indexing
> 
> > 
> 
> > Until we overcome the constraints on the sequence duration, copying the `RxData` for a given number of repetitions is a good idea. I would use a counter for the repetions as we could employ it directly for the generation of an ISMRMRD file.
> 
> 
> 
> Want me to include this in this PR?

Yes, it would be nice. Since this PR effects the way we interact with the receive data, I would like to have one dedicated update.

schote commented:
I could finally test the update on our end. A few more remaining questions:
- The decimation method (`ddc_method`) can be specified in `RxData`, so far the default is used. If the method should be adjustable, we should set it as defined in the acquisition parameters, when the `RxData` instances are created.
- The decimation rate parameter in the `AcquisitionParameter` interface is now redundant, correct? The way the decimation rate is calculated now is definetly the preferred one.
- The `dwell_time` attribute in `AcquisitionData` is redundant in favor of the new `RxData`
- The declaration of variables is inconsistent. Even if it is not done consistently by now, I would advocate not introducing any new abbreviations. Using `num_samples` or `num_points` instead of `num_pnts`.
- A default saving method for the new `RxData` is missing. The MRD export requires a labeled sequence (for trajectory information) or the MRD header file. Should we fully rely on MRD or implement an additional, simple storage format, e.g. based on h5 which is less informed than the MRD?
- I think a `__getitem__` method could help to improve access to averages, but the implementation can probalby be postponed to a future PR

Toreil commented:
Hey David, thanks for the feedback, ill address your comments in order:

1. Good point, that was oversight on my part. It's fixed now so that the ddc method defined in the acquisition parameters is used in RxData.
2. Correct, its not been removed yet since I was worried if I removed it I would have issues loading the state file when swapping between branches. Once the PR is ready to go it can be removed.
3. It's been removed. Lets also agree on some naming conventions since spcm_dwell_time and dwell_time are used interchangeably which made me nervous about deleting the correct references to dwell_time 
4.  Changed num_pnts to number of points, and agree on naming conventions, see previous point. 
5. I dont know. I'm happy for it to be MRD but my feeling is that it's rather complex to set up. I also don't think that the saving method should be defined in the RxData class (entirely?) since that data is usually in list form. The data can be extracted and we can keep saving as we used to do (kspace, imgs, metadata etc.) but this does lead to information loss. I'm open to suggestions
6. Again, the fact that the receive data is usually a list of RxData has stopped me from implementing it in RxData, it should probably be part of the acquisition_data class, no?

Toreil commented on src/console/utilities/sequences/tse/tse_3d.py:
```diff
@@ -376,7 +376,10 @@ def constructor(
-            seq.add_block(grad_ro, adc, label_pe1, label_pe2)
+            label_echo = pp.make_label(type='SET', label="ECO", value=int(echo_num + 1))
+            label_tr = pp.make_label(type='SET', label="REP", value=int(train_num + 1))
+            label_img = pp.make_label(type="INC", label="IMA", value=True)
```
Yeah, the docstring is outdated:
https://github.com/imr-framework/pypulseq/blob/master/src/pypulseq/supported_labels_rf_use.py

Toreil commented on src/console/spcm_control/acquisition_control.py:
```diff
@@ -236,8 +256,12 @@ def run(self) -> AcquisitionData:
-            if num_gates > 0:
-                self.post_processing(self.sequence.parameter, self.sequence.rx_phase_offset)
+            # Append the receive data with current scan data
+            scan_data = copy.deepcopy(self.rx_card.rx_data)
```
Was able to eliminate the need for the deep copy by copying the gate_data rather than just assigning the gate_data to RxData.raw_data.

schote commented on examples/acquisition_manager.py:
```diff
@@ -36,8 +36,10 @@
+
 fig, ax = plt.subplots(1, 1, figsize=(5, 5))
-_ = ax.plot(np.abs(data.raw.squeeze().T))
+_ = ax.plot(np.abs(data.receive_data.squeeze().T))
```
`data.receive_data` is a list, `squeeze()` does not work

schote commented on examples/acquisition_manager.py:
```diff
@@ -36,8 +36,10 @@
 
 # %%
 
+scan_data = [rx_data.receive_data.processed_data for rx_data in data.receive_data]
```
This variable is not used, is it just for demonstration purposes?

schote commented on src/console/interfaces/rx_data.py:
```diff
@@ -0,0 +1,133 @@
+    phase_offset: float
+    freq_offset: float
+
+    # Averages tracking
```
When I tested averaging, I found the attribute a bit unintuitive, this is why I left the comment. 
Theoretically averaging is possible using pypulseq and the acquisition parameter. Probably you would always use one or the other, but it would be great if they work hand in hand. What do you think about setting the `scan_number` to the average label from pulseq (if available, else 0)?
In case there are additional averages configures by the acquisition parameter we just increment the counter.
We could also just keep it as is and postpone this to another PR.

schote reviewed (changes requested):
I think there is a little error when plotting the data in the acquisition manager example. Beside that I could succesfully test the latest example and would be happy to approve this PR once the example is fixed.

I added another thought to the averaging counter in RxData - up to you if we keep it as is or change it within this PR.
